packages/fxa-event-broker/lib/notificationProcessor.ts

```diff
@@ -130,85 +87,92 @@ class ServiceNotificationProcessor {
     this.webhookService.stop();
   }
 
+  private async handleDeleteEvent(message: deleteSchema) {
+    const clientIds = await this.db.fetchClientIds(message.uid);
+    clientIds.forEach(async clientId => {
+      const topicName = this.topicPrefix + clientId;
+      const messageId = await this.pubsub
+        .topic(topicName)
+        .publishJSON({ event: message.event, uid: message.uid });
+      this.logger.debug('publishedMessage', { topicName, messageId });
+    });
+  }
+
+  private async handleLoginEvent(message: loginSchema) {
+    // Sync and some logins don't emit a clientId, so we have nothing to track
+    if (!message.clientId) {
+      this.logger.debug('unwantedMessage', {
+        message
+      });
+      return;
+    }
+    await this.db.storeLogin(message.uid, message.clientId);
+  }
+
+  private async handleSubscriptionEvent(message: subscriptionUpdateSchema) {
+    const clientIds = await this.db.fetchClientIds(message.uid);
+
+    // Split the product capabilities by clientId each capability goes to
+    const notifyClientIds: { [clientId: string]: string[] } = {};
+    message.productCapabilities
+      .map(item => item.split(':'))
+      .forEach(([clientId, capability]) => {
+        if (notifyClientIds[clientId]) {
+          notifyClientIds[clientId].push(capability);
+        } else {
+          notifyClientIds[clientId] = [capability];
+        }
+      });
+
+    const baseMessage = {
+      capabilities: [],
+      event: message.event,
+      isActive: message.isActive,
+      uid: message.uid
+    };
+
+    Object.entries(notifyClientIds)
+      .filter(([clientId]) => clientIds.includes(clientId))
+      .forEach(async ([clientId, capabilities]) => {
+        const topicName = this.topicPrefix + clientId;
+        const rpMessage = Object.assign(
+          {},
+          {
+            ...baseMessage,
+            capabilities
+          }
+        );
+
+        // TODO: Failures to publish due to missing queue should be Sentry reported.
+        const messageId = await this.pubsub.topic(topicName).publishJSON(rpMessage);
+        this.logger.debug('publishedMessage', { topicName, messageId });
+      });
+  }
+
   private async handleMessage(sqsMessage: SQS.Message) {
     const body = JSON.parse(sqsMessage.Body || '{}');
-    let message: baseMessageSchema;
-    try {
-      message = joi.attempt(body, BASE_MESSAGE_SCHEMA);
-    } catch (err) {
-      this.logger.error('badBaseMessage', { err });
+    const message = ServiceNotification.from(this.logger, body);
+    if (!message) {
+      // Anything that isn't a message we want
+      this.logger.debug('unwantedMessage', { message });
```

Comment:
I get that this is just refactoring existing code, but here it seems `message` will be `undefined` if a match isn't found, so we'll be logging `{ undefined }`. Is that the intent, or should `body` be logged instead? (I noticed that, in `handleLoginEvent`, when an unwanted message is logged, the actual message content is included.)

Reply:
Ah, nice catch! That is indeed a bug, I should've logged the body, not the undefined message.